diff --git a/.gitignore b/.gitignore
index 677c465..619c43b 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1 +1,21 @@
-.bundle
+.vagrant
+# Berksfile.lock
+*~
+*#
+.#*
+\#*#
+.*.sw[a-z]
+*.un~
+/cookbooks
+
+# Bundler
+# Gemfile.lock
+bin/*
+.bundle/*
+
+.kitchen/
+.kitchen.local.yml
+
+test/nodes
+
+Berksfile.lock
diff --git a/.kitchen.yml b/.kitchen.yml
new file mode 100644
index 0000000..461312e
--- /dev/null
+++ b/.kitchen.yml
@@ -0,0 +1,65 @@
+---
+driver:
+  name: vagrant
+  require_chef_omnibus: true
+
+platforms:
+  - name: ubuntu-16.04
+    driver_config:
+      box: bento/ubuntu-16.04
+  - name: debian-8
+    driver_config:
+      box: bento/debian-8.6
+
+suites:
+  - name: mariadb-10.1
+    run_list:
+      - recipe[mysqld::mariadb_repository]
+      - recipe[mysqld::default]
+    attributes:
+      mysqld:
+        db_install: 'mariadb'
+        root_password: 'abcdef01234567890'
+        repository:
+          mariadb:
+            version: '10.1'
+  - name: mariadb-10.2
+    run_list:
+      - recipe[mysqld::mariadb_repository]
+      - recipe[mysqld::default]
+    attributes:
+      mysqld:
+        db_install: 'mariadb'
+        root_password: 'abcdef01234567890'
+        repository:
+          mariadb:
+            version: '10.2'
+  - name: percona-5.6
+    run_list:
+      - recipe[mysqld::percona_repository]
+      - recipe[mysqld::default]
+    attributes:
+      mysqld:
+        db_install: 'percona'
+        root_password: 'abcdef01234567890'
+        repository:
+          percona:
+            version: '5.6'
+  - name: percona-5.7
+    run_list:
+      - recipe[mysqld::percona_repository]
+      - recipe[mysqld::default]
+    attributes:
+      mysqld:
+        db_install: 'percona'
+        root_password: 'abcdef01234567890'
+        repository:
+          percona:
+            version: '5.7'
+  - name: mysql-from-packages
+    run_list:
+      - recipe[mysqld::default]
+    attributes:
+      mysqld:
+        db_install: 'mysql'
+        root_password: 'abcdef01234567890'
diff --git a/.travis.yml b/.travis.yml
index 320d305..8a980ce 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -4,5 +4,5 @@ cache: bundler
 rvm:
 - 2.3.1
 script:
-- bundle exec rubocop
+- bundle exec cookstyle
 - cd spec/recipes && bundle exec rspec *.rb; cd -
diff --git a/Berksfile b/Berksfile
new file mode 100644
index 0000000..34fea21
--- /dev/null
+++ b/Berksfile
@@ -0,0 +1,3 @@
+source 'https://supermarket.chef.io'
+
+metadata
diff --git a/Gemfile b/Gemfile
index 56eb6d0..f3b3bbb 100644
--- a/Gemfile
+++ b/Gemfile
@@ -1,6 +1,7 @@
 source 'https://rubygems.org'
 
-gem 'rubocop'
+gem 'cookstyle'
+gem 'foodcritic'
 gem 'rspec'
 gem 'chefspec'
 gem 'berkshelf'
diff --git a/Gemfile.lock b/Gemfile.lock
index 4a84018..b8adc8b 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -3,6 +3,7 @@ GEM
   specs:
     addressable (2.4.0)
     ast (2.3.0)
+    backports (3.6.8)
     berkshelf (5.1.0)
       addressable (~> 2.3, >= 2.3.4)
       berkshelf-api-client (>= 2.0.2, < 4.0)
@@ -37,10 +38,10 @@ GEM
     celluloid-io (0.16.2)
       celluloid (>= 0.16.0)
       nio4r (>= 1.1.0)
-    chef (12.14.89)
+    chef (12.15.19)
       addressable
       bundler (>= 1.10)
-      chef-config (= 12.14.89)
+      chef-config (= 12.15.19)
       chef-zero (>= 4.8)
       diff-lcs (~> 1.2, >= 1.2.4)
       erubis (~> 2.7)
@@ -66,7 +67,7 @@ GEM
       specinfra (~> 2.10)
       syslog-logger (~> 1.6)
       uuidtools (~> 2.1.5)
-    chef-config (12.14.89)
+    chef-config (12.15.19)
       addressable
       fuzzyurl
       mixlib-config (~> 2.0)
@@ -77,21 +78,35 @@ GEM
       mixlib-log (~> 1.3)
       rack (~> 2.0)
       uuidtools (~> 2.1)
-    chefspec (5.2.0)
+    chefspec (5.3.0)
       chef (>= 12.0)
       fauxhai (~> 3.6)
       rspec (~> 3.0)
     cleanroom (1.0.0)
+    cookstyle (0.0.1)
+      rubocop (= 0.39.0)
+    cucumber-core (2.0.0)
+      backports (~> 3.6)
+      gherkin (~> 4.0)
     diff-lcs (1.2.5)
     erubis (2.7.0)
     faraday (0.9.2)
       multipart-post (>= 1.2, < 3)
-    fauxhai (3.9.0)
+    fauxhai (3.10.0)
       net-ssh
     ffi (1.9.14)
     ffi-yajl (2.3.0)
       libyajl2 (~> 1.2)
+    foodcritic (8.1.0)
+      cucumber-core (>= 1.3)
+      erubis
+      nokogiri (>= 1.5, < 2.0)
+      rake
+      rufus-lru (~> 1.0)
+      treetop (~> 1.4)
+      yajl-ruby (~> 1.1)
     fuzzyurl (0.9.0)
+    gherkin (4.0.0)
     hashie (3.4.6)
     highline (1.7.8)
     hitimes (1.2.4)
@@ -100,6 +115,7 @@ GEM
     ipaddress (0.8.3)
     json (2.0.2)
     libyajl2 (1.2.0)
+    mini_portile2 (2.1.0)
     minitar (0.5.4)
     mixlib-archive (0.2.0)
       mixlib-log
@@ -109,7 +125,7 @@ GEM
     mixlib-config (2.2.4)
     mixlib-log (1.7.1)
     mixlib-shellout (2.2.7)
-    molinillo (0.5.1)
+    molinillo (0.5.3)
     multi_json (1.12.1)
     multipart-post (2.0.0)
     net-scp (1.2.1)
@@ -124,9 +140,11 @@ GEM
       net-ssh-gateway (>= 1.2.0)
     net-telnet (0.1.1)
     nio4r (1.2.1)
-    octokit (4.3.0)
+    nokogiri (1.6.8.1)
+      mini_portile2 (~> 2.1.0)
+    octokit (4.4.1)
       sawyer (~> 0.7.0, >= 0.5.3)
-    ohai (8.20.0)
+    ohai (8.21.0)
       chef-config (>= 12.5.0.alpha.1, < 13)
       ffi (~> 1.9)
       ffi-yajl (~> 2.2)
@@ -141,10 +159,12 @@ GEM
     parser (2.3.1.4)
       ast (~> 2.2)
     plist (3.2.0)
+    polyglot (0.3.5)
     powerpack (0.1.1)
     proxifier (1.0.3)
     rack (2.0.1)
     rainbow (2.1.0)
+    rake (11.3.0)
     retryable (2.0.4)
     ridley (5.1.0)
       addressable
@@ -168,7 +188,7 @@ GEM
       rspec-core (~> 3.5.0)
       rspec-expectations (~> 3.5.0)
       rspec-mocks (~> 3.5.0)
-    rspec-core (3.5.3)
+    rspec-core (3.5.4)
       rspec-support (~> 3.5.0)
     rspec-expectations (3.5.0)
       diff-lcs (>= 1.2.0, < 2.0)
@@ -183,27 +203,28 @@ GEM
     rspec_junit_formatter (0.2.3)
       builder (< 4)
       rspec-core (>= 2, < 4, != 2.12.0)
-    rubocop (0.43.0)
-      parser (>= 2.3.1.1, < 3.0)
+    rubocop (0.39.0)
+      parser (>= 2.3.0.7, < 3.0)
       powerpack (~> 0.1)
       rainbow (>= 1.99.1, < 3.0)
       ruby-progressbar (~> 1.7)
       unicode-display_width (~> 1.0, >= 1.0.1)
     ruby-progressbar (1.8.1)
+    rufus-lru (1.1.0)
     sawyer (0.7.0)
       addressable (>= 2.3.5, < 2.5)
       faraday (~> 0.8, < 0.10)
     semverse (2.0.0)
-    serverspec (2.36.1)
+    serverspec (2.37.2)
       multi_json
       rspec (~> 3.0)
       rspec-its
       specinfra (~> 2.53)
-    sfl (2.2)
+    sfl (2.3)
     solve (3.0.1)
       molinillo (~> 0.4)
       semverse (>= 1.1, < 3.0)
-    specinfra (2.63.1)
+    specinfra (2.66.0)
       net-scp
       net-ssh (>= 2.7, < 4.0)
       net-telnet
@@ -213,12 +234,15 @@ GEM
     thor (0.19.1)
     timers (4.0.4)
       hitimes
+    treetop (1.6.8)
+      polyglot (~> 0.3)
     unicode-display_width (1.1.1)
     uuidtools (2.1.5)
     varia_model (0.6.0)
       buff-extensions (~> 2.0)
       hashie (>= 2.0.2, < 4.0.0)
     wmi-lite (1.0.0)
+    yajl-ruby (1.3.0)
 
 PLATFORMS
   ruby
@@ -226,8 +250,9 @@ PLATFORMS
 DEPENDENCIES
   berkshelf
   chefspec
+  cookstyle
+  foodcritic
   rspec
-  rubocop
 
 BUNDLED WITH
-   1.12.5
+   1.13.2
diff --git a/Rakefile b/Rakefile
new file mode 100644
index 0000000..3629294
--- /dev/null
+++ b/Rakefile
@@ -0,0 +1,26 @@
+#!/usr/bin/env rake
+
+require 'rspec/core/rake_task'
+RSpec::Core::RakeTask.new(:unit) do |t|
+  t.pattern = ['test/unit/**/*_spec.rb']
+end
+
+require 'foodcritic'
+FoodCritic::Rake::LintTask.new do |t|
+  t.options = { fail_tags: ['any'] }
+end
+
+require 'cookstyle'
+require 'rubocop/rake_task'
+RuboCop::RakeTask.new do |task|
+  task.options << '--display-cop-names'
+end
+
+begin
+  require 'kitchen/rake_tasks'
+  Kitchen::RakeTasks.new
+rescue LoadError
+  puts '>>>>> Kitchen gem not loaded, omitting tasks' unless ENV['CI']
+end
+
+task default: [:foodcritic, :rubocop]
diff --git a/attributes/defaults.rb b/attributes/defaults.rb
index 5b203b2..9784600 100644
--- a/attributes/defaults.rb
+++ b/attributes/defaults.rb
@@ -18,17 +18,38 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
 
-# Attribute that defines whether MariaDB or MySQL should be used
-default['mysqld']['use_mariadb'] = true
+# Attribute that defines whether MariaDB, MySQL or Percona should be used
+default['mysqld']['db_install'] = 'mariadb'
 
 # Default packages to install
-default['mysqld']['mysql_packages'] = %w(mysql-server)
-default['mysqld']['mariadb_packages'] = %w(mariadb-server)
-default['mysqld']['mariadb_galera_packages'] = %w(mariadb-galera-server)
+default['mysqld']['mysql']['packages'] = %w(mysql-server)
+default['mysqld']['mariadb']['packages'] = %w(mariadb-server)
+default['mysqld']['mariadb']['galera_packages'] = %w(mariadb-galera-server)
 
 # MariaDB repository configuration
-default['mysqld']['repository']['version'] = '10.1'
-default['mysqld']['repository']['mirror'] = 'http://ftp.hosteurope.de/mirror/mariadb.org/repo'
+default['mysqld']['repository']['mariadb']['version'] = '10.1'
+default['mysqld']['repository']['mariadb']['mirror'] = 'http://ftp.hosteurope.de/mirror/mariadb.org/repo'
+
+# Percona repository configuration
+default['mysqld']['repository']['percona']['version'] = '5.7'
+default['mysqld']['repository']['percona']['mirror'] = 'http://repo.percona.com/apt'
+
+default['mysqld']['percona']['packages'] = if node['mysqld']['repository']['percona']['version'] == '5.7'
+                                             %w(percona-server-server-5.7)
+                                           elsif node['mysqld']['repository']['percona']['version'] == '5.6'
+                                             %w(percona-server-server-5.6)
+                                           else
+                                             %w(percona-server-server)
+                                           end
+
+default['mysqld']['mysql']['version'] = if node['platform'] == 'debian' && node['platform_version'].to_i < 10
+                                          # At this time, stretch still uses mysql 5.6
+                                          5.6
+                                        elsif node['platform'] == 'ubuntu' && node['platform_version'].to_f < 16.04
+                                          5.6
+                                        else
+                                          5.7
+                                        end
 
 # Configure services
 default['mysqld']['my.cnf_path'] = '/etc/mysql/my.cnf'
@@ -42,7 +63,7 @@
 default['mysqld']['auth'] = '--defaults-file=/etc/mysql/debian.cnf'
 
 # Options, only set by default/ available on MariaDB
-if node['mysqld']['use_mariadb']
+if node['mysqld']['db_install'] == 'mariadb'
   # Charset options are only set on MariaDB by default
   default['mysqld']['my.cnf']['client']['default-character-set'] = 'utf8mb4'
   default['mysqld']['my.cnf']['mysql']['default-character-set'] = 'utf8mb4'
@@ -54,6 +75,24 @@
   default['mysqld']['my.cnf']['mysqld_safe']['skip_log_error'] = true
 end
 
+# Password columns for user passwords
+case node['mysqld']['db_install']
+when 'mariadb'
+  default['mysqld']['pwd_col'] = 'Password'
+when 'percona'
+  default['mysqld']['pwd_col'] = if node['mysqld']['repository']['percona']['version'] == '5.6'
+                                   'Password'
+                                 else
+                                   'authentication_string'
+                                 end
+when 'mysql'
+  default['mysqld']['pwd_col'] = if node['mysqld']['mysql']['version'] < 5.7
+                                   'Password'
+                                 else
+                                   'authentication_string'
+                                 end
+end
+
 default['mysqld']['my.cnf']['mysqld']['user'] = 'mysql'
 default['mysqld']['my.cnf']['mysqld']['port'] = 3306
 default['mysqld']['my.cnf']['mysqld']['lc-messages-dir'] = '/usr/share/mysql'
diff --git a/metadata.rb b/metadata.rb
index eb5fe32..4a2a2d3 100644
--- a/metadata.rb
+++ b/metadata.rb
@@ -5,4 +5,7 @@
 description      'Installs/Configures mysqld'
 long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
 version          '2.1.0'
+issues_url       'https://github.com/chr4-cookbooks/mysqld/issues'
+source_url       'https://github.com/chr4-cookbooks/mysqld/'
+
 depends          'apt'
diff --git a/providers/password.rb b/providers/password.rb
index 107812a..b8eff7e 100644
--- a/providers/password.rb
+++ b/providers/password.rb
@@ -21,7 +21,7 @@
 
 action :set do
   r = execute "Assign mysql password for #{new_resource.user} user" do
-    query = "UPDATE user SET authentication_string = PASSWORD('#{Shellwords.escape(new_resource.password)}') WHERE User = '#{Shellwords.escape(new_resource.user)}'"
+    query = "UPDATE user SET #{Shellwords.escape(node['mysqld']['pwd_col'])} = PASSWORD('#{Shellwords.escape(new_resource.password)}') WHERE User = '#{Shellwords.escape(new_resource.user)}'"
     command %(mysql #{Shellwords.escape(new_resource.auth)} mysql -e "#{query}; FLUSH PRIVILEGES;")
     only_if %(mysql #{Shellwords.escape(new_resource.auth)} -e 'SHOW DATABASES;')
   end
diff --git a/recipes/install.rb b/recipes/install.rb
index 7cb7168..a8aa0ca 100644
--- a/recipes/install.rb
+++ b/recipes/install.rb
@@ -18,9 +18,15 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
 
+# This way we force to have updated indexes
+include_recipe 'apt'
+
 # Install packages
-if node['mysqld']['use_mariadb']
-  Array(node['mysqld']['mariadb_packages']).each { |pkg| package pkg }
+case node['mysqld']['db_install']
+when 'mariadb'
+  Array(node['mysqld']['mariadb']['packages']).each { |pkg| package pkg }
+when 'percona'
+  Array(node['mysqld']['percona']['packages']).each { |pkg| package pkg }
 else
-  Array(node['mysqld']['mysql_packages']).each { |pkg| package pkg }
+  Array(node['mysqld']['mysql']['packages']).each { |pkg| package pkg }
 end
diff --git a/recipes/mariadb_galera_install.rb b/recipes/mariadb_galera_install.rb
index b387878..0a93799 100644
--- a/recipes/mariadb_galera_install.rb
+++ b/recipes/mariadb_galera_install.rb
@@ -19,4 +19,4 @@
 #
 
 # Install MariaDB + Galera packages
-Array(node['mysqld']['mariadb_galera_packages']).each { |pkg| package pkg }
+Array(node['mysqld']['mariadb']['galera_packages']).each { |pkg| package pkg }
diff --git a/recipes/mariadb_repository.rb b/recipes/mariadb_repository.rb
index e7db682..d0e63cf 100644
--- a/recipes/mariadb_repository.rb
+++ b/recipes/mariadb_repository.rb
@@ -21,12 +21,10 @@
 require 'uri'
 
 apt_repository 'mariadb' do
-  uri "#{node['mysqld']['repository']['mirror']}/#{node['mysqld']['repository']['version']}/#{node['platform']}"
+  uri "#{node['mysqld']['repository']['mariadb']['mirror']}/#{node['mysqld']['repository']['mariadb']['version']}/#{node['platform']}"
   distribution node['lsb']['codename']
   components %w(main)
   keyserver 'keyserver.ubuntu.com'
-
-  # Since Ubuntu xenial, the repository uses a new key
   if node['platform_version'].to_f < 16.04
     key '0xcbcb082a1bb943db'
   else
@@ -39,7 +37,7 @@
   mode 0o644
   content <<-EOS
     Package: *
-    Pin: origin #{URI.parse(node['mysqld']['repository']['mirror']).host}
+    Pin: origin #{URI.parse(node['mysqld']['repository']['mariadb']['mirror']).host}
     Pin-Priority: 1000
   EOS
 end
diff --git a/recipes/percona_repository.rb b/recipes/percona_repository.rb
new file mode 100644
index 0000000..00ccfa7
--- /dev/null
+++ b/recipes/percona_repository.rb
@@ -0,0 +1,38 @@
+#
+# Cookbook Name:: mysqld
+# Recipe:: percona_repository
+#
+# Copyright 2014, Chris Aumann
+# Copyright 2016, Adrian Almenar
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+apt_repository 'percona' do
+  uri node['mysqld']['repository']['percona']['mirror']
+  distribution node['lsb']['codename']
+  components %w(main)
+  keyserver 'keyserver.ubuntu.com'
+  key '8507EFA5'
+end
+
+# Prioritize Percona repository over system packages
+file '/etc/apt/preferences.d/percona.pref' do
+  mode 0o644
+  content <<-EOS
+    Package: *
+    Pin: release o=Percona Development Team
+    Pin-Priority: 1001
+  EOS
+end
diff --git a/spec/recipes/Berksfile.lock b/spec/recipes/Berksfile.lock
index 06418eb..97e90ec 100644
--- a/spec/recipes/Berksfile.lock
+++ b/spec/recipes/Berksfile.lock
@@ -7,5 +7,5 @@ GRAPH
   apt (4.0.2)
     compat_resource (>= 12.10)
   compat_resource (12.14.6)
-  mysqld (2.0.0)
+  mysqld (2.1.0)
     apt (>= 0.0.0)
diff --git a/spec/recipes/default_spec.rb b/spec/recipes/default_spec.rb
index 2f67c66..7743e01 100644
--- a/spec/recipes/default_spec.rb
+++ b/spec/recipes/default_spec.rb
@@ -8,12 +8,14 @@
   end
 
   let(:ubuntu_1604) do
-    ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '16.04').converge(described_recipe)
+    runner = ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '16.04')
+    runner.node.override['mysqld']['db_install'] = 'mysql'
+    runner.converge(described_recipe)
   end
 
   it 'should use the correct mysql server package' do
-    expect(debian.node['mysqld']['mysql_packages']).to eq(%w(mysql-server))
-    expect(ubuntu_1604.node['mysqld']['mysql_packages']).to eq(%w(mysql-server))
+    expect(debian.node['mysqld']['mysql']['packages']).to eq(%w(mysql-server))
+    expect(ubuntu_1604.node['mysqld']['mysql']['packages']).to eq(%w(mysql-server))
   end
 
   it 'should use distribution specific my.cnf path' do
@@ -38,4 +40,9 @@
     expect(ubuntu_1604.node['mysqld']['my.cnf']['mysqld']['myisam-recover-options']).to eq('BACKUP')
     expect(debian.node['mysqld']['my.cnf']['mysqld']['myisam-recover-options']).to eq('BACKUP')
   end
+
+  it 'should use the correct password column' do
+    expect(ubuntu_1604.node['mysqld']['pwd_col']).to eq('authentication_string')
+    expect(debian.node['mysqld']['pwd_col']).to eq('Password')
+  end
 end
diff --git a/test/integration/mariadb-10.1/serverspec/mariadb-10.1_spec.rb b/test/integration/mariadb-10.1/serverspec/mariadb-10.1_spec.rb
new file mode 100644
index 0000000..52230cf
--- /dev/null
+++ b/test/integration/mariadb-10.1/serverspec/mariadb-10.1_spec.rb
@@ -0,0 +1,23 @@
+require 'serverspec'
+require 'pathname'
+
+set :backend, :exec
+
+describe service('mysql') do
+  it { should be_enabled }
+  it { should be_running }
+end
+
+describe port(3306) do
+  it { should be_listening }
+end
+
+describe command('mysql -uroot -pabcdef01234567890 -N -s -r -e \'select version();\'') do
+  its(:stdout) { should match '^10\.1\.' }
+  its(:stdout) { should match 'MariaDB' }
+end
+
+describe command('mysql -udebian-sys-maint -pabcdef01234567890 -N -s -r -e \'select version();\'') do
+  its(:stdout) { should match '^10\.1\.' }
+  its(:stdout) { should match 'MariaDB' }
+end
diff --git a/test/integration/mariadb-10.2/serverspec/mariadb-10.2_spec.rb b/test/integration/mariadb-10.2/serverspec/mariadb-10.2_spec.rb
new file mode 100644
index 0000000..f896f67
--- /dev/null
+++ b/test/integration/mariadb-10.2/serverspec/mariadb-10.2_spec.rb
@@ -0,0 +1,23 @@
+require 'serverspec'
+require 'pathname'
+
+set :backend, :exec
+
+describe service('mysql') do
+  it { should be_enabled }
+  it { should be_running }
+end
+
+describe port(3306) do
+  it { should be_listening }
+end
+
+describe command('mysql -uroot -pabcdef01234567890 -N -s -r -e \'select version();\'') do
+  its(:stdout) { should match '^10\.2\.' }
+  its(:stdout) { should match 'MariaDB' }
+end
+
+describe command('mysql -udebian-sys-maint -pabcdef01234567890 -N -s -r -e \'select version();\'') do
+  its(:stdout) { should match '^10\.2\.' }
+  its(:stdout) { should match 'MariaDB' }
+end
diff --git a/test/integration/mysql-from-packages/serverspec/mysql-from-packages_spec.rb b/test/integration/mysql-from-packages/serverspec/mysql-from-packages_spec.rb
new file mode 100644
index 0000000..ae71301
--- /dev/null
+++ b/test/integration/mysql-from-packages/serverspec/mysql-from-packages_spec.rb
@@ -0,0 +1,21 @@
+require 'serverspec'
+require 'pathname'
+
+set :backend, :exec
+
+describe service('mysql') do
+  it { should be_enabled }
+  it { should be_running }
+end
+
+describe port(3306) do
+  it { should be_listening }
+end
+
+describe command('mysql -uroot -pabcdef01234567890 -N -s -r -e \'select version();\'') do
+  its(:stdout) { should match '^5\.' }
+end
+
+describe command('mysql -udebian-sys-maint -pabcdef01234567890 -N -s -r -e \'select version();\'') do
+  its(:stdout) { should match '^5\.' }
+end
diff --git a/test/integration/percona-5.6/serverspec/percona-5.6_spec.rb b/test/integration/percona-5.6/serverspec/percona-5.6_spec.rb
new file mode 100644
index 0000000..dfc4321
--- /dev/null
+++ b/test/integration/percona-5.6/serverspec/percona-5.6_spec.rb
@@ -0,0 +1,21 @@
+require 'serverspec'
+require 'pathname'
+
+set :backend, :exec
+
+describe service('mysql') do
+  it { should be_enabled }
+  it { should be_running }
+end
+
+describe port(3306) do
+  it { should be_listening }
+end
+
+describe command('mysql -uroot -pabcdef01234567890 -N -s -r -e \'select version();\'') do
+  its(:stdout) { should match '^5\.6\.' }
+end
+
+describe command('mysql -udebian-sys-maint -pabcdef01234567890 -N -s -r -e \'select version();\'') do
+  its(:stdout) { should match '^5\.6\.' }
+end
diff --git a/test/integration/percona-5.7/serverspec/percona-5.7_spec.rb b/test/integration/percona-5.7/serverspec/percona-5.7_spec.rb
new file mode 100644
index 0000000..c9b5cb9
--- /dev/null
+++ b/test/integration/percona-5.7/serverspec/percona-5.7_spec.rb
@@ -0,0 +1,17 @@
+require 'serverspec'
+require 'pathname'
+
+set :backend, :exec
+
+describe service('mysql') do
+  it { should be_enabled }
+  it { should be_running }
+end
+
+describe port(3306) do
+  it { should be_listening }
+end
+
+describe command('mysql -uroot -pabcdef01234567890 -N -s -r -e \'select version();\'') do
+  its(:stdout) { should match '^5\.7\.' }
+end
